design-docs/design-001.md

```diff
@@ -0,0 +1,236 @@
+- **PY_START** – Create a `Function` event for the code object and push a new frame ID onto the thread's stack.
+  ```rs
+  pub fn on_py_start(frame: *mut PyFrameObject);
+  ```
+- **PY_RESUME** – Emit an `Event` log noting resumption and update the current frame's state.
+  ```rs
+  pub fn on_py_resume(frame: *mut PyFrameObject);
+  ```
+- **PY_RETURN** – Pop the frame ID, write a `Return` event with the value (if retrievable), and link to the caller.
+  ```rs
+  pub struct ReturnRecord { pub frame: FrameId, pub value: Option<ValueRecord> }
+  pub fn on_py_return(frame: *mut PyFrameObject, value: *mut PyObject);
+  ```
+- **PY_YIELD** – Record a `Return` event flagged as a yield and keep the frame on the stack for later resumes.
+  ```rs
+  pub fn on_py_yield(frame: *mut PyFrameObject, value: *mut PyObject);
+  ```
+- **STOP_ITERATION** – Emit an `Event` indicating iteration exhaustion for the current frame.
+  ```rs
+  pub fn on_stop_iteration(frame: *mut PyFrameObject);
+  ```
+- **PY_UNWIND** – Mark the beginning of stack unwinding and note the target handler in an `Event`.
+  ```rs
+  pub fn on_py_unwind(frame: *mut PyFrameObject);
+  ```
+- **PY_THROW** – Emit an `Event` describing the thrown value and the target generator/coroutine.
+  ```rs
+  pub fn on_py_throw(frame: *mut PyFrameObject, value: *mut PyObject);
+  ```
+- **RERAISE** – Log a re-raise event referencing the original exception.
+  ```rs
+  pub fn on_reraise(frame: *mut PyFrameObject, exc: *mut PyObject);
+  ```
+
+### Call and Line Tracking
+- **CALL** – Record a `Call` event, capturing argument values and the callee's `Function` ID.
+  ```rs
+  pub fn on_call(callee: *mut PyObject, args: &PyTupleObject) -> FrameId;
+  ```
+- **LINE** – Write a `Step` event with current path and line number; ensure the path is registered.
+  ```rs
+  pub fn on_line(frame: *mut PyFrameObject, lineno: u32);
+  ```
+- **INSTRUCTION** – Optionally emit a fine-grained `Event` containing the opcode name for detailed traces.
+  ```rs
+  pub fn on_instruction(frame: *mut PyFrameObject, opcode: u8);
+  ```
+- **JUMP** – Append an `Event` describing the jump target offset for control-flow visualization.
+  ```rs
+  pub fn on_jump(frame: *mut PyFrameObject, target: u32);
+  ```
+- **BRANCH** – Record an `Event` with branch outcome (taken or not) to aid coverage analysis.
+  ```rs
+  pub fn on_branch(frame: *mut PyFrameObject, taken: bool);
+  ```
+
+### Exception Lifecycle
+- **RAISE** – Emit an `Event` containing exception type and message when raised.
+  ```rs
+  pub fn on_raise(frame: *mut PyFrameObject, exc: *mut PyObject);
+  ```
+- **EXCEPTION_HANDLED** – Log an `Event` marking when an exception is caught.
+  ```rs
+  pub fn on_exception_handled(frame: *mut PyFrameObject);
+  ```
+
+### C API Boundary
+- **C_RETURN** – On returning from a C function, emit a `Return` event tagged as foreign and include result summary.
+  ```rs
+  pub fn on_c_return(func: *mut PyObject, result: *mut PyObject);
```

Comment:
we can integrate this with rr backend in the future somehow or with a custom db-backend based C recording(??) but for now seems reasonable